tree/dataframe/inc/ROOT/RDFNodes.hxx

```diff
@@ -86,7 +86,27 @@ class RRangeBase;
 using RangeBasePtr_t = std::shared_ptr<RRangeBase>;
 using RangeBaseVec_t = std::vector<RangeBasePtr_t>;
 
-class RLoopManager {
+class RLoopManager;
+
+/// Base class for non-leaf nodes of the computational graph.
+/// It only exposes the bare minimum interface required to work as a generic part of the computation graph.
+/// RDataFrames and results of transformations can be cast to this type via ROOT::RDF::ToCommonNodeType.
+class RNode {
+protected:
+   RLoopManager *fLoopManager;
```

Comment:
Given that users will keep one of those around, how certain are you about them getting the lifetime of `RNode` vs its `fLoopManager` right? Does this need a `weak_ptr` here?

Reply:
Hi @Axel-Naumann , thank you for reviewing these changes!
Dataframe nodes are "internal" entities used when running the event loop.
Users only ever deal with nodes wrapped by `RInterface` (iow every user-handled node is always wrapped by a `RInterface`), which decorates them with all the user-facing methods and also with sanity checks such as the one you suggest.

`RInterface`, as well as `RResultPtr`s, do have a `weak_ptr<RLoopManager>` and guarantee that the graph that the user forms/executes is well formed. Nodes can always be sure that when they need to do stuff, the graph is well formed. I think when `fLoopManager` is used I often commented things like "we can be sure RLoopManager is alive here because we are _in the loop_", but I should probably add an explanation to the data member itself.

Reply:
P.S.
when i'll be done with #2049 , each node will own its parent, so we will always be sure that everything that needs to be in scope is, and branches of the computation graph will automagically go out of scope when they are not needed anymore